File: cpp/216.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> res;
        vector<int> temp;
        dfs(temp, res, 1, n, 9, k, 0);
        return res;
    }
    void dfs(vector<int> temp, vector<vector<int>> &res, int idx, int target, int n, int k, int depth)
    {

        if (depth == k) {
            if (target == 0) res.emplace_back(temp);
            return;
        }
        for (int i = idx; i <= n; ++i) {
                // printf("%d==%d, target-i:%d, i+1:%d \n", i, target, target-i, i+1);
            if (target-i >= 0) {  // 剪枝
                temp.push_back(i);
                dfs(temp, res, i+1, target-i, n, k, depth+1);
                temp.pop_back();
            }
        }
    }
};

File: python/216.py
from typing import List

class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        res = []
        temp = []
        candidates = list(range(1, 10))
        target = n
        n = len(candidates)
        
        def dfs(_target: int, idx:int, depth:int):
            if depth == k:
                if _target == 0:
                    res.append(temp[:])
                return
        
            for idx in range(idx, n):
                val = candidates[idx]
                if val <= _target:
                    temp.append(val)
                    dfs(_target - val, idx+1, depth+1)
                    temp.pop()
                    
                     
        dfs(target, 0, 0)
        return res